// @storybook/csf-plugin/dist/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.webpack = exports.vite = exports.unplugin = exports.rollup = exports.esbuild = void 0;

var _unplugin = require("unplugin");

var _promises = _interopRequireDefault(require("fs/promises"));

var _csfTools = require("@storybook/csf-tools");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const STORIES_REGEX = /\.(story|stories)\.[tj]sx?$/;
const logger = console;
const unplugin = (0, _unplugin.createUnplugin)(options => {
  return {
    name: 'unplugin-csf',
    enforce: 'pre',

    loadInclude(id) {
      return STORIES_REGEX.test(id);
    },

    async load(fname) {
      const code = await _promises.default.readFile(fname, 'utf-8');

      try {
        const csf = (0, _csfTools.loadCsf)(code, {
          makeTitle: userTitle => userTitle || 'default'
        }).parse();
        (0, _csfTools.enrichCsf)(csf);
        return (0, _csfTools.formatCsf)(csf);
      } catch (err) {
        logger.warn(err.message);
        return code;
      }
    }

  };
});
exports.unplugin = unplugin;
const {
  vite,
  rollup,
  webpack,
  esbuild
} = unplugin;
exports.esbuild = esbuild;
exports.webpack = webpack;
exports.rollup = rollup;
exports.vite = vite;
